### ISSUE TITLE
Rare partial upload issue with `consume_uploaded_entries` to external provider (S3)

### ISSUE THREAD

**Issue by petemcfarlane** [phoenixframework/phoenix_live_view#2037]
### Environment

* Elixir version (elixir -v): 1.13.3
* Phoenix version (mix deps): 1.6.9
* Phoenix LiveView version (mix deps): 0.17.7
* Operating system: alpine:3.15.3
* Browsers you attempted to reproduce this bug on (the more the merrier): not sure
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: yes

### Actual behavior

We're experiencing some rare but unexpected behaviour from using `consume_uploaded_entries` with LiveView.

```
** (ArgumentError) cannot consume uploaded files when entries are still in progress
    (phoenix_live_view 0.17.7) lib/phoenix_live_view/upload.ex:238: Phoenix.LiveView.Upload.consume_uploaded_entries/3
    (platform 0.0.1) lib/platform_web/live/us_qualifications_live/show.ex:43: PlatformWeb.USQualificationsLive.Show.handle_event/3
    (phoenix_live_view 0.17.7) lib/phoenix_live_view/channel.ex:349: anonymous fn/3 in Phoenix.LiveView.Channel.view_handle_event/3
    (telemetry 1.1.0) /root/platform/deps/telemetry/src/telemetry.erl:320: :telemetry.span/3
    (phoenix_live_view 0.17.7) lib/phoenix_live_view/channel.ex:206: Phoenix.LiveView.Channel.handle_info/2
    (stdlib 3.17.2) gen_server.erl:695: :gen_server.try_dispatch/4
    (stdlib 3.17.2) gen_server.erl:771: :gen_server.handle_msg/6
    (stdlib 3.17.2) proc_lib.erl:236: :proc_lib.wake_up/3
```

We've had several of these errors over several days just this week. We're uploading direct to S3. But we can't reproduce the issue ourselves.

We have a form which has a single file upload input

```elixir
def mount() do
  socket = 
      socket
      # set some other assigns here...
      |> allow_upload(:certificate,
        accept: ~w(.jpg .jpeg .png .pdf .gif),
        max_entries: 1,
        max_file_size: FileUploads.max_file_size(),
        external: &presign_upload/2
      )

    {:ok, socket}
end

defp presign_upload(entry, socket) do
  key = generate_key(socket.assigns.candidate_id)
  {:ok, fields} = FileUploads.make_policy(key, entry.client_type)

  meta = %{
    uploader: "S3",
    key: key,
    url: FileUploads.form_action(),
    fields: fields
  }

  {:ok, meta, socket}
end

def handle_event("save", params, socket) do
  uploaded_file =
    socket
    |> consume_uploaded_entries(:certificate, fn %{key: key}, _ ->
      {:ok, %{"key" => key, "bucket" => FileUploads.user_upload_bucket()}}
    end)
    # this is where it blows up!
end
```

In the Heex template
```elixir
<%= form_for @changeset, @path,
        [phx_change: "validate", phx_submit: "save", id: "form-id"],
        fn f -> %>


        <%= live_file_input @uploads.certificate, required: true %>
<% end %>
```

I found a few threads about this issue before on GitHub and in the Elixir Forum but they all seem like the issues were resolved in earlier versions of LiveView. 

Does anyone have any idea as to how we might try to recreate this issue, or the next steps to debug? Unfortunately, we don't currently have any logs about what the upload entry (or entries?) were, so I'm thinking we might try and catch the exception, and add some extra info to our logs. 

### Expected behavior

The exception should not be raised.

Any advice on what to try next would be greatly appreciated, thanks!

---

**EDIT 23rd May 2022**
I forgot to share the javascript uploader part, largely based on [these docs](https://hexdocs.pm/phoenix_live_view/uploads-external.html#direct-to-s3)
```javascript
Uploaders.S3 = (entries, onViewError) => {
  entries.forEach((entry) => {
    const formData = new FormData();
    const { url, fields } = entry.meta;
    Object.entries(fields).forEach(([key, val]) => formData.append(key, val));
    formData.append('file', entry.file);
    const xhr = new XMLHttpRequest();
    onViewError(() => xhr.abort());
    xhr.onload = () =>
      xhr.status === 201 ? entry.progress(100) : entry.error();
    xhr.onerror = () => entry.error();
    xhr.upload.addEventListener('progress', (event) => {
      if (event.lengthComputable) {
        const percent = Math.round((event.loaded / event.total) * 100);
        if (percent < 100) {
          entry.progress(percent);
        }
      }
    });

    xhr.open('POST', url, true);
    xhr.send(formData);
  });
};
```

**Comment by mcrumm** [comment]
Hi @petemcfarlane – the issue may be that you are using an older flavor of the form helper that doesn't play nicely with change tracking. What happens if you update your template to use the [`<.form>` function component](https://hexdocs.pm/phoenix_html/Phoenix.HTML.Form.html#module-phoenix-liveview-integration) style?

**Comment by petemcfarlane** [comment]
Hi @mcrumm thanks for that suggestion - I have since updated to use that - hopefully that could be it! The uploader has been in production for ~17months now, not that it gets a lot of use - but I think it must be working most of the time, just strange that we had this cluster of issues this week.

So, to help my understanding, are you suggesting that using the old `form_for` function with the `f -> ...` anonymous function, there might be some condition where the form fields aren't rendered to the DOM, therefore when the form submit event happens it isn't waiting for the uploading files to finish uploading? Sounds plausible if so!

**Comment by chrismccord** [comment]
I scanned to see if we could have a race b/w 100% progress and submitting the form, but I don't see one. I can't say what could be happening without knowing more. When a user submits the form, we block the form until all the uploads have reported 100% progress, then we trigger the save event, so this shouldn't be happening unless the client is able to submit the form button somehow. Are you by chance intercepting form submit or triggering the submit from JS rather than a standard button?

**Comment by petemcfarlane** [comment]
Thanks for the help @chrismccord. I've since updated the PR description to include our JS Uploaders, as you can see it's largely based on [the docs](https://hexdocs.pm/phoenix_live_view/uploads-external.html#direct-to-s3). We don't have anything that's intercepting a submit event, just using a `<%+ submit "Save" %>` button.

I'm going to try to expand on the [argument exception raised here](https://github.com/phoenixframework/phoenix_live_view/blob/07420030c7135b99aad4abb5f866dba88723f866/lib/phoenix_live_view/upload.ex#L238), to see what entry/entries are still being uploaded

```elixir
uploaded_file =
  try do
    socket
    |> consume_uploaded_entries(:certificate, fn %{key: key}, _ ->
      {:ok, %{"key" => key, "bucket" => FileUploads.user_upload_bucket()}}
    end)
  rescue
    e in ArgumentError ->
      require Logger
      Logger.info(inspect(uploaded_entries(socket, :certificate)))
      raise e
  end
```
Do you think something like that might help? Anything else you would try to debug?

**Comment by petemcfarlane** [comment]
Hi @chrismccord @mcrumm , I managed to recreate the issue in my app, and create a minimal reproducible scenario, based on the examples in HexDocs. 

[See loom video walkthrough here](https://www.loom.com/share/91c2de65c1a740f49f4e3cf237f5f6c6).

<details><summary>Click here to expand my LiveView module</summary>

```elixir
defmodule DemoWeb.FileUploadLive do
  use DemoWeb, :live_view

  @impl Phoenix.LiveView
  def mount(_params, _session, socket) do
    {:ok,
     socket
     |> assign(:uploaded_files, [])
     |> allow_upload(
       :avatar,
       accept: ~w(.jpg .jpeg),
       max_entries: 1,
       external: &presign_upload/2
     )}
  end

  @impl true
  def render(assigns) do
    ~H"""
    <form phx-change="validate" phx-submit="save">
      <%= live_file_input @uploads.avatar %>
      <%= for entry <- @uploads.avatar.entries do %>
        <%= for err <- upload_errors(@uploads.avatar, entry) do %>
          <p class="alert alert-danger"><%= error_to_string(err) %></p>
        <% end %>
      <% end %>
      <%= submit "Save" %>
    </form>
    """
  end

  @impl Phoenix.LiveView
  def handle_event("validate", _params, socket) do
    {:noreply, socket}
  end

  @impl Phoenix.LiveView
  def handle_event("save", _params, socket) do
    socket
    |> consume_uploaded_entries(:avatar, fn %{key: key}, _ ->
      {:ok, %{"key" => key, "bucket" => "bucket"}}
    end)

    {:noreply, socket}
  end

  defp presign_upload(entry, socket) do
    uploads = socket.assigns.uploads
    bucket = "phx-upload-example"
    key = "public/#{entry.client_name}"

    config = %{
      region: "us-east-1",
      access_key_id: "minioadmin",
      secret_access_key: "minioadmin"
    }

    {:ok, fields} =
      SimpleS3Upload.sign_form_upload(config, bucket,
        key: key,
        content_type: entry.client_type,
        max_file_size: uploads.avatar.max_file_size,
        expires_in: :timer.hours(1)
      )

    meta = %{
      uploader: "S3",
      key: key,
      url: "http://localhost:9000/#{bucket}",
      fields: fields
    }

    {:ok, meta, socket}
  end

  defp error_to_string(:too_large), do: "Too large"
  defp error_to_string(:too_many_files), do: "You have selected too many files"
  defp error_to_string(:not_accepted), do: "You have selected an unacceptable file type"
  defp error_to_string(:external_client_failure), do: "Cannot connect to external client"
end
```

</details>

<details>
<summary>Click to see simple_s3_upload.ex</summary>

```elixir
defmodule SimpleS3Upload do
  @moduledoc """
  Dependency-free S3 Form Upload using HTTP POST sigv4

  https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-post-example.html
  """

  @doc """
  Signs a form upload.

  The configuration is a map which must contain the following keys:

    * `:region` - The AWS region, such as "us-east-1"
    * `:access_key_id` - The AWS access key id
    * `:secret_access_key` - The AWS secret access key


  Returns a map of form fields to be used on the client via the JavaScript `FormData` API.

  ## Options

    * `:key` - The required key of the object to be uploaded.
    * `:max_file_size` - The required maximum allowed file size in bytes.
    * `:content_type` - The required MIME type of the file to be uploaded.
    * `:expires_in` - The required expiration time in milliseconds from now
      before the signed upload expires.

  ## Examples

      config = %{
        region: "us-east-1",
        access_key_id: System.fetch_env!("AWS_ACCESS_KEY_ID"),
        secret_access_key: System.fetch_env!("AWS_SECRET_ACCESS_KEY")
      }

      {:ok, fields} =
        SimpleS3Upload.sign_form_upload(config, "my-bucket",
          key: "public/my-file-name",
          content_type: "image/png",
          max_file_size: 10_000,
          expires_in: :timer.hours(1)
        )

  """
  def sign_form_upload(config, bucket, opts) do
    key = Keyword.fetch!(opts, :key)
    max_file_size = Keyword.fetch!(opts, :max_file_size)
    content_type = Keyword.fetch!(opts, :content_type)
    expires_in = Keyword.fetch!(opts, :expires_in)

    expires_at = DateTime.add(DateTime.utc_now(), expires_in, :millisecond)
    amz_date = amz_date(expires_at)
    credential = credential(config, expires_at)

    encoded_policy =
      Base.encode64("""
      {
        "expiration": "#{DateTime.to_iso8601(expires_at)}",
        "conditions": [
          {"bucket":  "#{bucket}"},
          ["eq", "$key", "#{key}"],
          {"acl": "public-read"},
          ["eq", "$Content-Type", "#{content_type}"],
          ["content-length-range", 0, #{max_file_size}],
          {"x-amz-server-side-encryption": "AES256"},
          {"x-amz-credential": "#{credential}"},
          {"x-amz-algorithm": "AWS4-HMAC-SHA256"},
          {"x-amz-date": "#{amz_date}"}
        ]
      }
      """)

    fields = %{
      "key" => key,
      "acl" => "public-read",
      "content-type" => content_type,
      "x-amz-server-side-encryption" => "AES256",
      "x-amz-credential" => credential,
      "x-amz-algorithm" => "AWS4-HMAC-SHA256",
      "x-amz-date" => amz_date,
      "policy" => encoded_policy,
      "x-amz-signature" => signature(config, expires_at, encoded_policy)
    }

    {:ok, fields}
  end

  defp amz_date(time) do
    time
    |> NaiveDateTime.to_iso8601()
    |> String.split(".")
    |> List.first()
    |> String.replace("-", "")
    |> String.replace(":", "")
    |> Kernel.<>("Z")
  end

  defp credential(%{} = config, %DateTime{} = expires_at) do
    "#{config.access_key_id}/#{short_date(expires_at)}/#{config.region}/s3/aws4_request"
  end

  defp signature(config, %DateTime{} = expires_at, encoded_policy) do
    config
    |> signing_key(expires_at, "s3")
    |> sha256(encoded_policy)
    |> Base.encode16(case: :lower)
  end

  defp signing_key(%{} = config, %DateTime{} = expires_at, service) when service in ["s3"] do
    amz_date = short_date(expires_at)
    %{secret_access_key: secret, region: region} = config

    ("AWS4" <> secret)
    |> sha256(amz_date)
    |> sha256(region)
    |> sha256(service)
    |> sha256("aws4_request")
  end

  defp short_date(%DateTime{} = expires_at) do
    expires_at
    |> amz_date()
    |> String.slice(0..7)
  end

  defp sha256(secret, msg), do: :crypto.mac(:hmac, :sha256, secret, msg)
end

```
</details>

In summary:
1. Set `allow_upload` options `max_entries: 1`, and set `external` to some `presign_upload` function, based on the HexDocs example.
2. If there is some error from the client uploading the file (e.g. some network issue) we see a client error (I simulate this by making up a server, e.g. using http://localhost:9000)
3. Change the file input file selection and the `save` event is re-triggered, even before the new file has been uploaded.

I hope this information helps with the debugging, and thank you for your efforts!

(PS sorry for the weird background grunting noise in the video, that's my 2mo old daughter making a guest appearance 👶 )

**Comment by mcrumm** [comment]
Thanks for the video and for digging deeper into this @petemcfarlane! I was able to reproduce the issue: mcrumm/live_upload_example@f3f7ef2d8873bb167f0baff8cc2f5e4ae71e3890

You can work around the problem by first canceling the failed upload: mcrumm/live_upload_example@1f9dc77b47b8b4b06a30fcf54c0d6e83e65767fb

However the submit being triggered after selecting a new file looks like a bug. That sort of thing should only occur with `auto_upload: true`.
